### PR TITLE
Lock detect etc cleanup

### PR DESCRIPTION
**Pull request by henryhallam**
Cleaning up the lock detector stuff.

This version does not disable RAIM.

### CODE REVIEW

**Comment by henryhallam** [comment]
@fnoble @gsmcmullin @cbeighley Please could you review? This is doing well on HITL, slightly outperforming v0.19-rc0.  In addition to the lock detector stuff, there are two important bug fixes: the TOW extraction in libswiftnav:nav_msg.c and the NMEA string buffer safety.

File: src/acq.c [hunk 1]
```diff
@@ -183,15 +183,17 @@ void acq_service_irq(void)
...
-void acq_get_results(float* cp, float* cf, float* snr)
+void acq_get_results(float* cp, float* cf, float* cn0)
 {
   *cp = 1023.0 - (float)(acq_state.best_cp % (1023 * NAP_ACQ_CODE_PHASE_UNITS_PER_CHIP))
                   / NAP_ACQ_CODE_PHASE_UNITS_PER_CHIP;
   *cf = (float)acq_state.best_cf / NAP_ACQ_CARRIER_FREQ_UNITS_PER_HZ;
   /* "SNR" estimated by peak power over mean power. */
-  *snr = (float)acq_state.best_power / (acq_state.power_acc / acq_state.count);
+  float snr = (float)acq_state.best_power / (acq_state.power_acc / acq_state.count);
+  *cn0 = 10 * log10(snr)
```
Comment (denniszollo) [inline src/acq.c]: Thoughts on trapping a NAN return here with a LOG_ERROR?  I don't really want 0 SNR and NAN cn0 floating around?

Comment (cbeighley) [inline src/acq.c]: +1 add a check for `snr == 0` and `log_error` if so

File: src/manage.c [hunk 2]
```diff
@@ -323,17 +324,14 @@ static void manage_acq()
     return;
   }
 
-  log_info("acq: PRN %d found @ %d Hz, %d SNR\n", prn + 1, (int)cf, (int)snr);
```
Comment (cbeighley) [inline src/manage.c]: Added an issue to piksi_tools to replace this print with a view: https://github.com/swift-nav/piksi_tools/issues/124

File: src/manage.c [hunk 3]
```diff
@@ -370,7 +367,7 @@ static u8 manage_track_new_acq(void)
   return MANAGE_NO_CHANNELS_FREE;
 }
 
-/** Clear unhealthy flags after some time.  Flags are reset one per day. */
+/** Clear unhealthy flags after some time.  Flags are reset once per day. */
```
Comment (denniszollo) [inline src/manage.c]: This comment is ambigous.  Should say reset by the airforce or government.

I read this as the frequency that the firmware calls this function (which is every 5 times through the code, not oncer per day)

File: src/manage.c [hunk 4]
```diff
@@ -404,26 +401,11 @@ static msg_t manage_track_thread(void *arg)
...
-
 void manage_track_setup()
 {
   initialize_lock_counters();
 
-  SETTING_NOTIFY("track", "iq_output_mask", iq_output_mask, TYPE_INT,
-                 track_iq_output_notify);
-
-  SETTING("track", "track_cn0_threshold", track_cn0_threshold, TYPE_FLOAT);
+  SETTING("track", "cn0_use", track_cn0_use_thres, TYPE_FLOAT);
```
Comment (denniszollo) [inline src/manage.c]: Lets make sure to put a corresponding commit to settings.yaml in piksi tools with a documentation string

Comment (henryhallam) [inline src/manage.c]: yeah, there is

File: src/acq.c (hunk 1 above)
Comment (henryhallam) [inline src/acq.c]: ok will do

File: src/manage.c (hunk 3 above)
Comment (henryhallam) [inline src/manage.c]: No, the firmware is clearing the flags once per day (see a couple lines further down).  I can clarify the comment further.

File: src/manage.c (hunk 4 above)
Comment (cbeighley) [inline src/manage.c]: Have you updated `settings.yaml` in `piksi_tools` to document these new / changed settings?

File: src/track.c [hunk 5]
```diff
@@ -395,8 +462,12 @@ void tracking_channel_update(u8 channel)
       break;
     }
     case TRACKING_DISABLED:
+      /* TODO: Why do we sometimes have to disable the channel in the
+               NAP repeatedly? */
```
Comment (cbeighley) [inline src/track.c]: Add `log_warn` here?

Comment (henryhallam) [inline src/track.c]: Yeah I tried that but it was spamming every time a tracking channel got disabled.  I think it's because the FPGA may still do one more update.

File: src/manage.c [hunk 6]
```diff
@@ -433,84 +415,105 @@ void manage_track_setup()
...
+
+    /* Optimistic phase lock detector "unlocked" for a while? */
+    /* TODO: This isn't doing much.  Use the pessimistic detector instead? */
+    if ((int)(uc - ch->ld_opti_locked_count) > TRACK_DROP_UNLOCKED_T) {
+      log_info("PRN%d PLL unlocked too long, dropping\n", ch->prn+1);
+      drop_channel(i);
+      continue;
+    }
+
+#define CHECKSENTINEL(X) for (int j=0; j < SENTINEL_WORDS_EACH; j++) if (X[j] != 0x12345678) {\
```
Comment (cbeighley) [inline src/manage.c]: #define for `0x12345678`?

File: src/track.h [hunk 7]
```diff
@@ -29,15 +29,24 @@
 
 extern u8 n_rollovers;
 
+#define SENTINEL_WORDS_EACH 1
```
Comment (cbeighley) [inline src/track.h]: Sentinel is for memory corruption testing? Add some comments about it? Do you think we should keep it in here now that the issue has been ostensibly resolved?

File: src/track.c [hunk 8]
```diff
@@ -358,21 +418,22 @@ void tracking_channel_update(u8 channel)
...
-                 chan->prn+1, (unsigned int)chan->update_count);
+        log_info("PRN %d synced @ %u ms, %.1f dBHz\n",
+                 chan->prn+1, (unsigned int)chan->update_count, chan->cn0);
         chan->stage = 1;
         struct loop_params *l = &loop_params_stage[1];
         chan->int_ms = l->coherent_ms;
         chan->short_cycle = true;
 
-        /* TODO What is BW for C/N0 estimation? */
-        cn0_est_init(&chan->cn0_est, 1e3 / l->coherent_ms, chan->cn0, 5,
```
Comment (cbeighley) [inline src/track.c]: Use #define for 5?

Comment (cbeighley) [inline src/track.c]: Oops, nevermind.

File: src/track.c [hunk 9]
```diff
@@ -343,14 +402,15 @@ void tracking_channel_update(u8 channel)
         * NAP_TRACK_CARRIER_FREQ_UNITS_PER_HZ;
 
 #if 1
-      if (chan->int_ms != 1) {
+      if (chan->stage > 0 && chan->lock_detect.outo) {
         s32 I = (cs[1].I - chan->alias_detect.first_I) / (chan->int_ms - 1);
         s32 Q = (cs[1].Q - chan->alias_detect.first_Q) / (chan->int_ms - 1);
         float err = alias_detect_second(&chan->alias_detect, I, Q);
         if (fabs(err) > (250 / chan->int_ms)) {
```
Comment (cbeighley) [inline src/track.c]: #define for 250?

File: src/track.c [hunk 10]
```diff
@@ -343,14 +402,15 @@ void tracking_channel_update(u8 channel)
         * NAP_TRACK_CARRIER_FREQ_UNITS_PER_HZ;
 
 #if 1
```
Comment (cbeighley) [inline src/track.c]: Any reason to keep the conditional compilation here?

File: src/track.h (hunk 7 above)
Comment (henryhallam) [inline src/track.h]: Yes.  I haven't seen any instances of it in 100+ HITL runs, so I'll remove it.  I might make a separate PR to do more general checking of memory getting written where it shouldn't.

File: src/track.c (hunk 10 above)
Comment (henryhallam) [inline src/track.c]: I should probably make a setting to enable/disable false lock detection.

File: src/track.c (hunk 9 above)
Comment (henryhallam) [inline src/track.c]: I'm not totally clear on where it comes from.. need to do a bit more work to understand / validate the alias detection / correction stuff.  That can be a separate PR/issue though, it was in master before this PR.

File: src/track.h (hunk 7 above)
Comment (cbeighley) [inline src/track.h]: Sounds good.

File: src/track.c (hunk 9 above)
Comment (cbeighley) [inline src/track.c]: Sounds good.

**Comment by henryhallam** [comment]
Removed sentinels and rebased

**Comment by cbeighley** [comment]
lgtm
